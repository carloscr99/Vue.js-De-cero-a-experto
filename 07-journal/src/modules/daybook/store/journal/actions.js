import journalApi from "../../../../api/journalApi"


// export const myAction = async({ commit }) =>{

// }
    

export const loadEntries = async( { commit } ) =>{

    const { data } = await journalApi.get('/entries.json')

    if( !data ){
        commit('setEntries', [])
        return
    }

    const entries = []
    for( let id of Object.keys(data)){
        console.log(id)
        entries.push({
            id,
            ...data[id]
        })
    }

    // Con el commit llamamos a las mutations, y enviamos las entradas
    commit('setEntries', entries)
    
}
export const updateEntry = async( {commit} , entry) => { //entry debe de ser un parametro

    console.log('updateEntry -> commit: ', commit)

    console.log('updateEntry -> entry: ', entry)
    
    // Extraer solo la información necesaria (quitar el ID)
    const { date, picture, text } = entry

    const dataToSave =  { date, picture, text }

    const resp = await journalApi.put( `entries/${entry.id}.json`, dataToSave)

    console.log(resp)

    //Commit de una mutación -> updateEntry
    //... para no enviar el objeto como referencia
    commit('updateEntry', {...entry})
    
}
export const createEntry = async( { commit }, entry ) =>{

    const { date, picture, text } = entry
    const dataToSave =  { date, picture, text }


    const {data} = await journalApi.post( `/entries.json`, dataToSave)
    
    dataToSave.id = data.name
    
    commit('addEntry', dataToSave)

    return data.name

}

export const deleteEntry = async({ commit }, id) =>{

     await journalApi.delete(`/entries/${id}.json`)

    commit('deleteEntry', id) 

}