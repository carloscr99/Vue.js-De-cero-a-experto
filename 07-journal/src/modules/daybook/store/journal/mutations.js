

// export const myMutations = ( state ) =>{
    
// }

// Las mutaciones son las que pueden cambiar el state

export const setEntries = (  state, entries ) =>{

    state.entries = [ ...state.entries, ...entries ]
    state.isLoading = false
    
}
export const updateEntry = ( /* state */ ) =>{
    
}
export const addEntry = ( /* state */ ) =>{
    
}