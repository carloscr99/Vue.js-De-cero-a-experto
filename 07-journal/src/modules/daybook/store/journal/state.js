

export default () => ({
    isLoading: true,
    entries: [
        {
        id: '1',
        date: new Date().toDateString(),
        text: 'Et mollit ullamco sit aliqua dolore incididunt eu do deserunt laborum fugiat. Duis laborum dolore nisi officia. Tempor culpa sunt et Lorem velit amet non magna non proident quis velit. Culpa adipisicing laborum enim excepteur nostrud in elit labore cillum excepteur enim cupidatat sit. Do nisi adipisicing ea veniam id ex sint id cillum.',
        picture: null,
        
    },
    {
        id: '2',
        date: new Date().toDateString(),
        text: 'Consequat qui esse cillum laboris sunt tempor nisi aute cillum. Proident ad ea cillum elit Lorem eiusmod quis reprehenderit ut aliquip amet qui. Lorem cupidatat culpa velit adipisicing mollit. Ipsum labore quis occaecat occaecat fugiat. Consequat pariatur tempor Lorem culpa aliqua est magna adipisicing ea ex do adipisicing proident et. Id commodo aliqua irure qui.',
        picture: null,
        
    },
    {
        id: '3',
        date: new Date().toDateString(),
        text: 'Exercitation aliquip laboris do laborum veniam ut sunt. Elit pariatur occaecat laboris proident magna magna eu. Deserunt nulla in labore dolore. Aliquip ipsum enim tempor id labore nisi velit est anim. Excepteur occaecat sint deserunt ipsum sint esse.x',
        picture: null,
        
    },
]
})