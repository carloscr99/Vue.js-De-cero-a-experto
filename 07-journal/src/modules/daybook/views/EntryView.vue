<template>
    <template v-if="entry">
        <div class="entry-title d-flex justify-content-between p-2">

            <div>
                <span class="text-success fs-3 fw-bold">{{day}}</span>
                <span class="mx-1 fs-3">{{month}}</span>
                <span class="mx-2 fs-4 fw-light">{{ yearDay }}</span>
            </div>

            <div>
                <button class="btn btn-danger mx-2">
                    Borrar
                    <i class="fa fa-trash-alt"></i>
                </button>
                <button class="btn btn-primay">
                    Subir foto
                    <i class="fa fa-upload"></i>
                </button>
            </div>

        </div>
        <hr>
        <div class="d-flex flex-column px-3 h-75">

            <textarea v-model="entry.text" placeholder="¿Que tal el día?"></textarea>

        </div>
        <Fab icon="fa-save"/>

        <img src="https://external-content.duckduckgo.com/iu/?u=https%3A%2F%2Ftse1.mm.bing.net%2Fth%3Fid%3DOIP.OciRQXR70LOSpcj15y-xagHaEb%26pid%3DApi&f=1" alt="entry-picture"
        class="img-thumbnail">


    </template>
</template>


<script>
import { defineAsyncComponent } from 'vue';
import { mapGetters } from 'vuex';  // computed!
import getDayMonthYear from '../helpers/getDayMonthYear'

export default {
    props: {
        id: {
            type: String,
            required: true
        }
    },
    components: {
        Fab: defineAsyncComponent(() =>import('../components/Fab.vue'))
    },
    //Objeto reactivo
    data() {
        return {
            entry: null
        }
    },

    computed: {
        ...mapGetters('journal', ['getEntryById']),
        day() {
            const { day } = getDayMonthYear( this.entry.date )
            return day
        },
        month() {
            const { month } = getDayMonthYear( this.entry.date )
            return month
        },
        yearDay() {
            const { yearDay } = getDayMonthYear( this.entry.date )
            return yearDay
        }
    },
    methods: {
        loadEntry(){
            const entry = this.getEntryById(this.id)
            if ( !entry ) this.$router.push({ name: 'no-entry' })

            this.entry = entry
        }
    },
    created(){
        console.log(this.id)
        this.loadEntry()
    },
    //Cada vez que el id cambia, cargamos la entrada correspondiente
    watch: {
        id( ){
           this.loadEntry() 
        }
    }
}
</script>


<style lang="scss" scoped>
textarea {
    font-size: 20px;
    border: none;
    height: 100%;

    &:focus{
        outline: none;
    }
}

img {
    width: 200px;
    position: fixed;
    bottom: 150px;
    right: 20px;
    box-shadow: 0px 5px 10px rgba($color: #000000, $alpha: 0.2);
}

</style>