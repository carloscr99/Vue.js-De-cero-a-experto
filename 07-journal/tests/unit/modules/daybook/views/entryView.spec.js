import { shallowMount } from "@vue/test-utils"
import { createStore } from "vuex"
import { journalState } from "../../../mock-data/test-journal-state"
import journal from '@/modules/daybook/store/journal'
import EntryView from '@/modules/daybook/views/EntryView.vue'
import Swal from 'sweetalert2'

const createViewStore = ( initialState ) =>
createStore({
    modules: {
        journal: {
            //Desestructuramos todo el contenido
            ...journal,
            state: {
                ...initialState
            }
        }
    }
})

jest.mock('sweetalert2', () =>({
    fire: jest.fn(),
    showLoading: jest.fn(),
    close: jest.fn()

}))

describe('Pruebas en el entryView', () =>{

    const store = createViewStore(journalState)

    store.dispatch = jest.fn()
    
    const mockRouter = {
        push: jest.fn()
    }

   

    let wrapper 

    beforeEach(()=>{
        jest.clearAllMocks()
        wrapper = shallowMount(EntryView, {
            props: {
                id: '-Mti8otNoP8i_GJlRjei',
            },
            global: {
                mocks: {
                    $router: mockRouter
                },
                plugins: [ store ]
            }
        })
    
    })




    test('debe de sacar al usuario porque el id no existe', ()=>{

        const wrapper = shallowMount(EntryView, {
            props: {
            //id: '-Mti8otNoP8i_GJlRjei',
            id: 'id falso',
            },
            global: {
                mocks: {
                    $router: mockRouter
                },
                plugins: [ store ]
            }
        })


        expect( mockRouter.push ).toHaveBeenCalledWith({ name: 'no-entry' })
        
        
    })
    
    test('debe de mostrar la entrada correctamente', ()=>{
        
        
        expect(wrapper.html()).toMatchSnapshot()
        expect( mockRouter.push ).not.toHaveBeenCalled()


    })


    test('debe de borrar la entrada y salir',  ()=>{

        Swal.fire.mockReturnValueOnce( Promise.resolve({ isConfirmed: true }))

        wrapper.find('.btn-danger').trigger('click')


        expect (Swal.fire).toHaveBeenCalledWith({
            title: '¿Estás seguro?',
            text: 'Una vez borrado, no se puede recuperar',
            icon: 'warning',
            showDenyButton: true,
            confirmButtonText: 'Si, estoy seguro'
        })

        setTimeout(()=>{
            expect(store.dispatch).toHaveBeenCalledWith('journal/deleteEntry', '-Mti8otNoP8i_GJlRjei')
            expect(mockRouter.push).toHaveBeenCalled()
            done()

        }, 1)

    })

})