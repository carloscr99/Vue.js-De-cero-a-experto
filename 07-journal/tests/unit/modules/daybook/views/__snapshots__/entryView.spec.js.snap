// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en el entryView debe de mostrar la entrada correctamente 1`] = `
<div class="entry-title d-flex justify-content-between p-2">
  <div><span class="text-success fs-3 fw-bold">18</span><span class="mx-1 fs-3">Enero</span><span class="mx-2 fs-4 fw-light">2022, Martes</span></div>
  <div><input type="file" accept="image/png, image/jpeg" style="display: none;"><button class="btn btn-danger mx-2"> Borrar <i class="fa fa-trash-alt"></i></button><button class="btn btn-primay"> Subir foto <i class="fa fa-upload"></i></button></div>
</div>
<hr>
<div class="d-flex flex-column px-3 h-75"><textarea placeholder="¿Que tal el día?"></textarea></div>
<fab-stub icon="fa-save"></fab-stub>
<img src="https://res.cloudinary.com/dh7vxoobp/image/upload/v1642525586/curso-vue/pozoaza6dalj76fxwic6.png" alt="entry-picture" class="img-thumbnail">
<!--v-if-->
`;
